instance/models/openedx_instance.py

```diff
@@ -249,54 +251,104 @@ def delete(self, *args, **kwargs):
         studio_domain = tldextract(self.internal_studio_domain)
         gandi.set_dns_record(
             studio_domain.registered_domain,
-            type='CNAME', name=studio_domain.subdomain, value=lms_domain.subdomain
+            type='CNAME', name=studio_domain.subdomain, value=load_balancer_domain
         )
 
+    def get_load_balancer_configuration(self):
+        """
+        Return the haproxy configuration fragment and backend map for this instance.
+        """
+        if self.active_appserver:
+            backend_name = "be-{}".format(self.active_appserver.server.name)
+            server_name = "appserver-{}".format(self.pk)
+            ip_address = self.active_appserver.server.public_ip
+        else:
+            backend_name = "be-preliminary-page-{}".format(self.pk)
+            server_name = "preliminary-page"
+            ip_address = settings.PRELIMINARY_PAGE_SERVER_IP
+        template = loader.get_template("instance/haproxy/backend.conf")
+        if ip_address is not None:
+            config = template.render(dict(
+                domain=self.domain,
+                backend_name=backend_name,
+                server_name=server_name,
+                ip_address=ip_address,
+            ))
+        else:
+            # No active appserver and PRELIMINARY_PAGE_SERVER_IP not set, so there is no backend
+            # we can configure.  Simply return an empty configuration in this case.
+            self.logger.info("PRELIMINARY_PAGE_SERVER_IP is unset, so not configuring a preliminary backend.")
+            return "", ""
+        domain_names = [
+            self.external_lms_domain, self.external_lms_preview_domain, self.external_studio_domain,
+            self.internal_lms_domain, self.internal_lms_preview_domain, self.internal_studio_domain,
+        ]
+        backend_map = "\n".join(" ".join([domain, backend_name]) for domain in domain_names if domain)
+        return backend_map, config
+
+    @property
+    def appserver_set(self):
+        """
+        Get the set of OpenEdxAppServers owned by this instance.
+        """
+        return self.ref.openedxappserver_set
+
+    def set_appserver_active(self, appserver_id):
+        """
+        Mark the AppServer with the given ID as the active one.
+        """
+        app_server = self.appserver_set.get(pk=appserver_id)  # Make sure the AppServer is owned by this instance
+        self.logger.info('Making %s active for instance %s...', app_server.name, self.name)
         self.active_appserver = app_server
         self.save()
-
+        self.load_balancing_server.reconfigure()
         self.enable_monitoring()
 
+    def set_appserver_inactive(self):
+        """
+        Reset the active AppServer to None.  Called when the active appserver is terminated.
+        """
+        self.disable_monitoring()
+        self.active_appserver = None
+        self.save()
+        self.load_balancing_server.reconfigure()
```

Comment:
Might be worth syncing up with @itsjeyd on this, since he's adding shutdown logic in https://github.com/open-craft/opencraft/pull/143/files#diff-9842c8f4a538c1c1735bdb68328d427aR84